### PR TITLE
feat: [PL-41523]: Create stepId to taskId mapping

### PR DESCRIPTION
**Pull request by mdmilic**
## Describe your changes
* Create stepId -> taskId mapping
* Separated bijou task scheduling into a separate GRPC class (out of DelegateServiceGrpcImpl as they have nothing in common)
* re-look all the protos to make sure they follow protobuf naming conventions
* Re-implemented K8SInfra mapper to do full supported mapping (we had only basic implementation before). Used mapstruct for mappings
* Removed logging token from the ingress API (now generated by DMS). 

AppSec tickets:
https://harness.atlassian.net/browse/SECAPP-1208
https://harness.atlassian.net/browse/SECAPP-1210

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53608)

### CODE REVIEW

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger uts

**Review by puthrayaharness — APPROVED** [review]

**Comment by mdmilic** [comment]
trigger uts

**Review by AnupamIO — APPROVED** [review]

File: 420-delegate-service/src/main/java/io/harness/grpc/scheduler/ScheduleTaskServiceGrpcImpl.java
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/grpc/scheduler/ScheduleTaskServiceGrpcImpl.java:1]: Thanks for doing this work !

File: 420-delegate-service/src/main/java/io/harness/delegate/resources/core/CoreDelegateExecutionResource.java [hunk 1]
```diff
@@ -98,41 +99,48 @@ public Response acquireRequestPayload(@PathParam("executionId") final String exe
...
       if (Objects.isNull(task)) {
-        log.error("When processing response from delegate {}, task not found.", delegateId);
+        log.error("Task not found when processing infra setup response from delegate {}", delegateId);
         return Response.serverError().build();
       }
-      final String uuid =
-          executionInfrastructureService.addExecutionInfrastructure(task, delegateId, response.getLocation());
 
-      delegateTaskService.handleResponse(task, taskQuery,
+      infraService.updateDelegateInfo(task.getUuid(), delegateId, response.getLocation().getDelegateName());
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/delegate/resources/core/CoreDelegateExecutionResource.java:133]: need to check the return ?  if the infra doesn't exist, should we fail the task right away

File: 420-delegate-service/src/main/java/io/harness/executionInfra/ExecutionInfraLocation.java [hunk 2]
```diff
@@ -29,10 +34,13 @@
 @Entity(value = "executionInfraLocation", noClassnameStored = true)
 @HarnessEntity(exportable = false)
 @FieldNameConstants(innerTypeName = "ExecutionInfraLocationKeys")
-public class ExecutionInfraLocation implements PersistentEntity, UuidAware, CreatedAtAware {
+public class ExecutionInfraLocation implements PersistentEntity, UuidAware, CreatedAtAware, UpdatedAtAware {
   @Setter @Id private String uuid;
   private final String delegateGroupName;
   private final String runnerType;
+  private final Map<String, String> stepTaskIds;
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/executionInfra/ExecutionInfraLocation.java:41]: it'd be very helpful if we add comment to this, what is key what is value

File: 920-delegate-service-beans/src/main/proto/io/harness/delegate/security_context.proto [hunk 3]
```diff
@@ -16,13 +16,13 @@ option go_package = "920_delegate_service_beans";
     privileged = false; runAsUser = 0 (root); runAsNonRoot = false;
  */
 message SecurityContext {
-  bool allowPrivilegeEscalation = 1;
+  bool allow_privilege_escalation = 1;
```
Comment (xingchi-jin) [inline 920-delegate-service-beans/src/main/proto/io/harness/delegate/security_context.proto:19]: are these all implemented in k8s runner ?

File: 420-delegate-service/src/main/java/io/harness/grpc/scheduler/ScheduleTaskServiceGrpcImpl.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+      } else {
+        log.error("Unexpected error occurred while processing submit task request.", ex);
+      }
+      responseObserver.onError(io.grpc.Status.INTERNAL.withDescription(ex.getMessage()).asRuntimeException());
+    }
+  }
+
+  private SetupExecutionInfrastructureResponse sendInitK8SInfraTask(
+      final SetupExecutionInfrastructureRequest request, final SchedulingConfig schedulingConfig) throws IOException {
+    final var loggingToken = getLoggingToken(request.getConfig().getAccountId());
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/grpc/scheduler/ScheduleTaskServiceGrpcImpl.java:130]: If call log service to get token using the same account id for multiple times, will we get the same token everytime ? subsequent question is, shall we cache it

File: 910-delegate-service-driver/src/main/java/io/harness/grpc/DelegateServiceDriverGrpcClientModule.java [hunk 5]
```diff
@@ -125,6 +133,13 @@ CallCredentials dsCallCredentials() {
     return new ServiceAuthCallCredentials(serviceSecret, new ServiceTokenGenerator(), "delegate-service");
   }
 
+  @Named("st-call-credentials")
+  @Provides
+  @Singleton
+  CallCredentials stCallCredentials() {
+    return new ServiceAuthCallCredentials(serviceSecret, new ServiceTokenGenerator(), "schedule-task-service");
```
Comment (xingchi-jin) [inline 910-delegate-service-driver/src/main/java/io/harness/grpc/DelegateServiceDriverGrpcClientModule.java:140]: i m not following this change, what is this for ?

File: 420-delegate-service/src/main/java/io/harness/grpc/scheduler/mapper/K8sInfraMapper.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    final var securityContext = SecurityContextMapper.map(spec.getSecurityContext());
+    final var compute = ComputeMapper.INSTANCE.map(spec.getComputeResource());
+
+    return k8SInfraBuilder.setCompute(compute)
+        .setSecurityContext(securityContext)
+        .setWorkingDir("/opt/harness")
+        .setLogPrefix(request.getInfra().getLogConfig().getLogKey())
+        .setLogToken(loggingToken)
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/grpc/scheduler/mapper/K8sInfraMapper.java:41]: can you help remove this line in this PR https://github.com/harness/harness-core/blob/develop/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/acquire_task.proto#L42C8-L42C8 

Previously i added this, but now i think it makes more sense to put log token in infra_data

**Review by mdmilic — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53608)* status: Some of the checks are too old. 51 file(s) to review. 5. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`420-delegate-service/src/main/java/io/harness/delegate/resources/core/CoreDelegateExecutionResource.java` line 133 at r5](https://reviewable.io/reviews/harness/harness-core/53608#gh-1346526747:-Ng4KtCj0GrKuxTLrIaG:b-qvshpd) ([raw file](https://github.com/harness/harness-core/blob/11a2220740a83bed0173f16c08899f1fbf743e0e/420-delegate-service/src/main/java/io/harness/delegate/resources/core/CoreDelegateExecutionResource.java#L133)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

need to check the return ?  if the infra doesn't exist, should we fail the task right away 
</blockquote></details>

Do you mean if the LocationRef record doesn't exist in Mongo? I think we should rather make sure it always exist at this point, I'll see that the InitInfra API fails if you don't provide the appropriate infra (i.e. fail fast). 
___
*[`420-delegate-service/src/main/java/io/harness/executionInfra/ExecutionInfraLocation.java` line 41 at r5](https://reviewable.io/reviews/harness/harness-core/53608#gh-1346527329:-Ng4LSwE9_dwXU9BZ6H5:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/11a2220740a83bed0173f16c08899f1fbf743e0e/420-delegate-service/src/main/java/io/harness/executionInfra/ExecutionInfraLocation.java#L41)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

it'd be very helpful if we add comment to this, what is key what is value 
</blockquote></details>

Done.
___
*[`420-delegate-service/src/main/java/io/harness/grpc/scheduler/ScheduleTaskServiceGrpcImpl.java` line 130 at r5](https://reviewable.io/reviews/harness/harness-core/53608#gh-1346548595:-Ng4LWsu5kWvJUIoAqFD:b-ml4q7p) ([raw file](https://github.com/harness/harness-core/blob/11a2220740a83bed0173f16c08899f1fbf743e0e/420-delegate-service/src/main/java/io/harness/grpc/scheduler/ScheduleTaskServiceGrpcImpl.java#L130)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

If call log service to get token using the same account id for multiple times, will we get the same token everytime ? subsequent question is, shall we cache it 
</blockquote></details>

We do have caching mechanism implemented, but that's part of a different PR waiting to be merged :smile:. After the other one gets merged, I'll update the code here to use cache, rather than the logging service client directly.
___
*[`420-delegate-service/src/main/java/io/harness/grpc/scheduler/mapper/K8sInfraMapper.java` line 41 at r5](https://reviewable.io/reviews/harness/harness-core/53608#gh-1346556690:-Ng4LyZb6L3wPKtY6CiZ:b-6y82jj) ([raw file](https://github.com/harness/harness-core/blob/11a2220740a83bed0173f16c08899f1fbf743e0e/420-delegate-service/src/main/java/io/harness/grpc/scheduler/mapper/K8sInfraMapper.java#L41)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

can you help remove this line in this PR https://github.com/harness/harness-core/blob/develop/955-delegate-beans/src/main/proto/io/harness/delegate/core/beans/acquire_task.proto#L42C8-L42C8 

Previously i added this, but now i think it makes more sense to put log token in infra_data
</blockquote></details>

Will do, TBH I'm undecided yet where is the best place for that logging config. K8SInfra is runner specific, yet this config is global (i.e. we must have it for each runner). On the other hand each runner will need a separate implementation of how it deals with the logging token so there is no commonalities apart from the fact that each runner would need this. We are currently using it from the K8S Infra and the one in TaskPayload is unused. 
___
*[`910-delegate-service-driver/src/main/java/io/harness/grpc/DelegateServiceDriverGrpcClientModule.java` line 140 at r5](https://reviewable.io/reviews/harness/harness-core/53608#gh-1346553911:-Ng4N7C4AU8BfcWtiMFv:b-y897nt) ([raw file](https://github.com/harness/harness-core/blob/11a2220740a83bed0173f16c08899f1fbf743e0e/910-delegate-service-driver/src/main/java/io/harness/grpc/DelegateServiceDriverGrpcClientModule.java#L140)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

i m not following this change, what is this for ? 
</blockquote></details>

I've extracted all the Bijou GRPC APIs (ingress APIs) into a separate "GRPC Service" (out of DelegateServiceGrpcImpl, because it wasn't really related at all to it - not reusing anything).
This will still be inside CG Manager (not a separate microservice), but goes over a separate communication channel (not the one for DelegateServiceGrpcImpl). We already have 3-4 of these (DelegateServiceGrpcImpl, DelegateProfileServiceGrpcImpl, DelegateDetailsServiceGrpcImpl)
___
*[`920-delegate-service-beans/src/main/proto/io/harness/delegate/security_context.proto` line 19 at r5](https://reviewable.io/reviews/harness/harness-core/53608#gh-1346532642:-Ng4NuZnAwzlZT8eVMiT:b2nhswy) ([raw file](https://github.com/harness/harness-core/blob/11a2220740a83bed0173f16c08899f1fbf743e0e/920-delegate-service-beans/src/main/proto/io/harness/delegate/security_context.proto#L19)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

are these all implemented in k8s runner ? 
</blockquote></details>

Yes

**Review by xingchi-jin — APPROVED** [review]
> Reviewed 26 of 49 files at r1, 8 of 14 files at r2, 4 of 6 files at r3, 2 of 4 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53608)* status: Some of the checks are too old. 51 file(s) to review. 

___
*[`420-delegate-service/src/main/java/io/harness/delegate/resources/core/CoreDelegateExecutionResource.java` line 133 at r5](https://reviewable.io/reviews/harness/harness-core/53608#gh-1346526747:-Ng6NE1v9Mak7cmNBtw5:b-2c4d25) ([raw file](https://github.com/harness/harness-core/blob/11a2220740a83bed0173f16c08899f1fbf743e0e/420-delegate-service/src/main/java/io/harness/delegate/resources/core/CoreDelegateExecutionResource.java#L133)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Do you mean if the LocationRef record doesn't exist in Mongo? I think we should rather make sure it always exist at this point, I'll see that the InitInfra API fails if you don't provide the appropriate infra (i.e. fail fast).
</blockquote></details>

yes, i mean entry disappear from Mongo due to TTL. ignore if this is not a concern 
___
*[`420-delegate-service/src/main/java/io/harness/grpc/scheduler/mapper/K8sInfraMapper.java` line 41 at r5](https://reviewable.io/reviews/harness/harness-core/53608#gh-1346556690:-Ng6NYyOE9dyVu6cJoc5:b-mg3s1g) ([raw file](https://github.com/harness/harness-core/blob/11a2220740a83bed0173f16c08899f1fbf743e0e/420-delegate-service/src/main/java/io/harness/grpc/scheduler/mapper/K8sInfraMapper.java#L41)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Will do, TBH I'm undecided yet where is the best place for that logging config. K8SInfra is runner specific, yet this config is global (i.e. we must have it for each runner). On the other hand each runner will need a separate implementation of how it deals with the logging token so there is no commonalities apart from the fact that each runner would need this. We are currently using it from the K8S Infra and the one in TaskPayload is unused.
</blockquote></details>

yep, currently we have the freedom to change APIs, when the first adoption in production, we will not have such freedom.
I m OK putting log config in either global place or in infra. Technically it's something infra uses, so it can put in both k8s infra, or vm runner infra of the future

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger uts

**Comment by mdmilic** [comment]
trigger ut0

**Review by AnupamIO — APPROVED** [review]

**Review by jyotiraval — APPROVED** [review]

**Review by puthrayaharness — APPROVED** [review]

**Pull request merged**